### ISSUE TITLE
Issue with the output of LDs - Forced Duffing

### ISSUE THREAD

**Issue by VikJGG** [broncio123/lds#5]
The current implementation of the code is giving an issue when computing the LD scalar field for the forced Duffing system, because the manifolds appear to be distorted. In the code, the trajectories of all initial conditions on the grid are constructed for the full time interval of the integration, and afterwards LDs is calculated using the complete trajectories stored in memory. This is done in the function below. 

* Could you please explain how the 'points_final' variable is organized internally in the code in order to produce the 'points_difference' variable that contains each line segment of the trajectory?

* What is the advantage of saving all the trajectory of the initial condition and then accumulate the LD value instead of accumulating at each time step of the integration? 

``` python
def lagrangian_descriptor(points_final, p_norm = 0.5, flag_m = True):
    points_difference = points_final[1:] - points_final[:-1]
    if flag_m: # p-norm LDs
        LD_rows = np.sum(np.abs(points_difference)**p_norm, axis=2)
        LD = np.sum(LD_rows, axis=0)
    else:  # arclength LDs
        LD_rows = np.sqrt(np.sum((points_final)**2),axis=2)
        LD = np.sum(LD_rows, axis=0)
    return LD
```

**Comment by broncio123** [comment]
Hi @VikJGG 

* For the first point, the distortion in the LD maps for Duffing has to do with the current implementation of the integrator `scipy.integrate.solve_ivp`. 

_CURRENT IMPLEMENTATION_

![image](https://user-images.githubusercontent.com/17453074/83396275-a347e600-a3f3-11ea-9d98-b0d23d2d905f.png)

_PREVIOUS IMPLEMENTATION(S)_

![image](https://user-images.githubusercontent.com/17453074/83396246-97f4ba80-a3f3-11ea-8889-58e5f43c69d6.png)

__EXPLANATION__

For the first, under `compute_lagrangian_descriptor` the `solution_object` is given by

```python
solution_object = integrator(f, time_interval, y0, t_eval=t_evaluation, args=(vector_field, ))
```
where 
```python
t_evaluation = np.arange(t_initial, t_final + dt, dt) 
```
is the sequence of time-points for the integration and at which the computed solution is stored.
See documentation [here](https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.solve_ivp.html#scipy.integrate.solve_ivp)

Since `t_evaluation` provides the time points to perform the integration, then `dt` the _timestep in the integration becomes fixed_ for the whole `time_interval = (t0, t0 + tau)` - taken as the `t_span` input of the integrator.

This, in turn, affects the accuracy of the computed trajectory, which seems to result in the distortion observed in the LD map.

The above implementation was added as a way to avoid using a `for` loop for the `time_interval` -  as done in previous implementations. However, doing so stops the integrator from using an adaptative timestep - which could be shorter than`dt`, and that may be necessary for some trajectories to achieve a decent resolution. In previous implementations, `t_span` was defined as `(t, t+dt)`, with some time `t` in the `time_interval`

* For the second point, `points_final` in the current implementation is a  1D `numpy` array whose entries are arrays of size `(Nx * Ny, 2)`, this array is a snapshot of all the trajectories at a common time `t` in the `time_interval`, for the whole `(Nx, Ny)` grid of `initial_points` in 2D. 

So, once you take `points_difference`, this will be another 1D `numpy` array, whose entries are again `(Nx * Ny, 2)` arrays, where each subentry will be the difference between a 2D-point at time `t+dt` and the previous one at time `t` for every trajectory for all `initial_points`

* PENDING ...

**Comment by vkrajnak** [comment]
Why should the integration timestep become fixed? The _solve_ivp_ still has to satisfy _atol_ and _rtol_, which guarantees precision. There may be arbitrarily many integration steps between the times specified in _t_evaluation_, which are times at which values are saved in the output object.

**Comment by vkrajnak** [comment]
@VikJGG I don't think it is very good to calculate trajectories first and then calculate LDs, but it is cleaner and more readable than the previous for loop in time and repeated initialisation of _solve_ivp_

_points_final_ is just a matrix of all trajectories at the times specified in _t_eval_ as returned by _solve_ivp_

**Comment by vkrajnak** [comment]
@broncio123 This issue should be resolved with my pull request. _solve_ivp_ seemingly applies _atol_ and _rtol_ per component and for the Duffing oscillator the predefined value wasn't enough.

@VikJGG LDs are calculated along the trajectories now, so the piece of code you mention is no longer in.
